cats_tutorial.ipynb

```diff
@@ -0,0 +1,491 @@
+    "\n",
+    "We have all of the information we need to choose a temperature for a specific room and time of day. To use VW to achieve this, we will do the following:\n",
+    "\n",
+    "We convert our context into the text format we need\n",
+    "We pass this example to vw and get the chosen action and the probability of chosing that action\n",
+    "Finally we return the temperature chosen, and the probability of choosing it (we are going to need the probability when we learn form this example)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 61,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def predict_temperature(vw, context):\n",
+    "    vw_text_example = to_vw_example_format(context)\n",
+    "    return vw.predict(vw_text_example)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Simulation set up\n",
+    "\n",
+    "Now that we have done all of the setup work and know how to interface with VW, let's simulate the world of our two rooms. The scenario is that the thermostat it turned on in each room and it has to propose a temperature. Remember that the reward function allows us to define the worlds reaction to what VW recommends.\n",
+    "\n",
+    "We will choose between Living Room and Bedroom uniformly at random and also choose the time of day uniformly at random. You can think of this as us tossing a coin to choose between the rooms (Living Room if heads and Bedroom if tails) and another coin toss for choosing time of day."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 62,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "rooms = ['Living Room', 'Bedroom']\n",
+    "times_of_day = ['morning', 'afternoon']\n",
+    "\n",
+    "def choose_room(rooms):\n",
+    "    return random.choice(rooms)\n",
+    "\n",
+    "def choose_time_of_day(times_of_day):\n",
+    "    return random.choice(times_of_day)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will instantiate a CB learner in VW and then simulate the thermostat interaction num_iterations number of times. In each interaction, we:\n",
+    "\n",
+    "1. Decide between Living and Bedroom\n",
+    "2. Decide time of day\n",
+    "3. Pass context i.e. (room, time of day) to learner to get a temperature i.e. a value between min (0 degrees) and max (32 degrees) and probability of choosing that temperature\n",
+    "4. Receive reward i.e. see if the proposed temperature was adjusted or not, and by how much. Remember that cost is just negative reward.\n",
+    "5. Format context, action (temperature), probability, reward in VW format\n",
+    "6. Learn from the example\n",
+    "\n",
+    "This is the same for every one of our simulations, so we define the process in the run_simulation function. The cost function must be supplied as this is essentially us simulating how the world works."
```

Comment:
The above steps are repeatedly executed during our simulations, so we define the process in the run_simulator function. ...